src/listeners/block-listener.service.ts

```diff
@@ -0,0 +1,219 @@
+      index: index.toHuman(),
+      data: rawData.toHuman(),
+      phase: phase.toHuman(),
+    };
+
+    return result;
+  }
+
+  private parseExtrinsicRecord(rawRecord, index: number): ExtrinsicData {
+    const {
+      method: { section, method },
+      isSigned,
+    } = rawRecord.toHuman();
+
+    const result = {
+      index,
+      section,
+      method,
+      isSigned,
+      signer: isSigned ? rawRecord.signer.toString() : null,
+      // args: JSON.stringify(rawRecord.args), // todo: Do we really need these args? Alternative way to get args: rawRecord.method.toJSON().args
+      hash: rawRecord.hash.toHex(),
+    };
+
+    return result;
+  }
+
+  private async getBlockDataByNumber(blockNumber: number): Promise<BlockData> {
+    const blockHash = await this.api.rpc.chain.getBlockHash(blockNumber);
+    return this.getBlockDataByHash(blockHash);
+  }
+
+  private async getBlockDataByHash(blockHash: BlockHash): Promise<BlockData> {
+    const [rawRuntimeVersion, rawBlock, apiAt] = await Promise.all([
+      this.api.runtimeVersion,
+      this.api.rpc.chain.getBlock(blockHash),
+      this.api.at(blockHash),
+    ]);
+
+    const [rawEvents, rawTotalIssuance] = await Promise.all([
+      apiAt.query.system.events(),
+      apiAt.query.balances.totalIssuance(),
+    ]);
+
+    const { header: rawHeader, extrinsics: rawExtrinsics } = rawBlock.block;
+
+    const timestamp = this.getTimestampFromExtrinsics(rawExtrinsics);
+
+    const parsedEvents = (rawEvents as unknown as Array<unknown>).map(
+      this.parseEventRecord.bind(this),
+    );
+
+    const parsedExtrinsics = (rawExtrinsics as unknown as Array<unknown>).map(
+      this.parseExtrinsicRecord.bind(this),
+    );
+
+    const result = {
+      blockNumber: rawHeader.number.toNumber(),
+      timestamp,
+      blockHash: blockHash.toHuman(),
+      extrinsics: parsedExtrinsics,
+      events: parsedEvents,
+      totalIssuance: rawTotalIssuance.toString(),
+      ...pick(rawHeader.toHuman(), [
+        'stateRoot',
+        'extrinsicsRoot',
+        'parentHash',
+      ]),
+      ...pick(rawRuntimeVersion.toJSON(), ['specName', 'specVersion']),
+    };
+
+    return result;
+  }
+
+  async startListening(): Promise<Observable<BlockData>> {
+    // Check api for readiness.
+    await this.apiService.isReady;
+
+    return new Observable((subscriber) => {
+      this.api.rpc.chain.subscribeNewHeads(async (lastHeader) => {
+        try {
+          const blockNumber = lastHeader.number.toNumber();
+
+          this.logger.verbose(
+            `Got new block ${blockNumber}`,
+            'BlockListenerService',
+          );
+
+          const result = await this.getBlockDataByHash(lastHeader.hash);
+
+          subscriber.next(result);
+        } catch (err) {
+          subscriber.error(err?.message);
+        }
+      });
+    });
+  }
+
+  async getBlockByNumber(blockNumber: number): Promise<Observable<BlockData>> {
```

Comment:
Something strange happening in this place.

Reply:
Ok. Now returning `Promise<BlockData>`